woocommerce/woocommerce-ios
# Add product from image: update text field to support AI-generated suggestion

Part of #10180 


## Why


We're running an A/B experiment for a new starting screen for the product creation flow. This new screen just contains 3 fields: image, name, and description. When selecting a packaging image, the texts can be scanned and used to generate product details (name and description for now). The latest POC looks like pe5sF9-1Es-p2#comment-2495.

The focus of this PR: the text fields now support multiple lines, and a suggestion box is shown if the text field is non-empty. When generating product details from the image, a loading UI is also shown to indicate that a suggestion might be available.

🗒️ There is a known issue where the multiline text field can have an unexpectedly taller height after applying the suggested text. It's a subtask in https://github.com/woocommerce/woocommerce-ios/issues/10180.

## How

- A new view `AddProductFromImageTextFieldView` was created for the name and description fields in the form `AddProductFromImageView`. Its view model can be invoked with a suggestion `onSuggestion`, and then conditionally updates either its `text` or `suggestedText` based on the following logic:
  - If the suggestion is empty or the same as the `text`, then it resets the `suggestedText`
  - If `text` is non-empty, the suggestion is set to `suggestedText`. Otherwise, the suggestion is set to `text` 
  - Test cases were added for this behavior
- `AddProductFromImageViewModel` and `AddProductFromImageView` were updated with the new text field view

🗒️ I've tried using a `TextEditor` for the multiline text field, but the height also has unexpected issues. Using a `TextField` with iOS 16+ multiline support has fewer height issues from my testing. Please feel free to suggest any improvements on the text field.

## Testing instructions


- Log in to a WPCOM store
- Go to the Products tab
- Tap + to add a product
- If the store has 0-2 products, tap `Add manually`
- Tap to select `Simple physical product`
- Enter some text to either the name or description field
- Tap on the image header to add an image with some text using any source (camera/device library/media library) --> after a list of the scanned texts is shown, there should be a loading state below both text fields. Shortly, if the details generation succeeds, the suggested text should be shown in a suggestion box for the non-empty text field
- Feel free to unselect and edit the text of some rows and test the regeneration flow
- Tap `Continue` --> the new screen should be dismissed and then it navigates to the main product form
- Tap `Publish` to publish the product --> the product should be saved remotely with the selected image and generated name and description

## Screenshots



https://github.com/woocommerce/woocommerce-ios/assets/1945542/cc4a732d-e24b-4605-b8e0-acbb8c49f549

light - loading | light - suggested text | dark - loading | dark - suggested text
-- | -- | -- | --
![Simulator Screenshot - iPhone 14 - 2023-07-13 at 13 59 06](https://github.com/woocommerce/woocommerce-ios/assets/1945542/4bc0e126-26ad-4afe-b4df-7e5388620d95) | ![Simulator Screenshot - iPhone 14 - 2023-07-13 at 13 59 18](https://github.com/woocommerce/woocommerce-ios/assets/1945542/a983a800-69f4-40ea-b4a5-e6a03a5320a8) | ![Simulator Screenshot - iPhone 14 - 2023-07-13 at 13 59 49](https://github.com/woocommerce/woocommerce-ios/assets/1945542/2737cd87-9a00-48ee-8e0f-d22503ba4dc2) | ![Simulator Screenshot - iPhone 14 - 2023-07-13 at 14 00 00](https://github.com/woocommerce/woocommerce-ios/assets/1945542/33ef67b3-063b-41be-b9c1-c70065b8a41f)


---
